### PR TITLE
Pass `--detect-features` to binaryen unless finalize has extracted them already

### PR DESCRIPTION
**Pull request by sbc100**
This means that running finalize at all can become optional.

### CODE REVIEW

**Comment by sbc100** [comment]
Adding @aardappel because it seems like we don't know how `memory64` ends up in the features sections.. so I can't see how finalize `--detect-features` would work in `memory64` mode.. since it would fail to validate I imagine.

**Comment by aardappel** [comment]
I don't think we need to put Memory64 in the features section, it be better if it simply looks at the memory section to see if the i64 index flag is set.

**Comment by sbc100** [comment]
> I don't think we need to put Memory64 in the features section, it be better if it simply looks at the memory section to see if the i64 index flag is set.

Either way I don't think this PR changes that.  Either binaryen can detect and it work.. or it can't and it doesn't.. both before and after this PR that should be true.

**Comment by sbc100** [comment]
> I don't think we need to put Memory64 in the features section, it be better if it simply looks at the memory section to see if the i64 index flag is set.

It would be easy to detect threads too, by the presence of shared memory, but we have not historically done that for features.. instead we have always asked users to opt-in to a given feature explicitly (or validation will fail).

File: emscripten.py [hunk 1]
```diff
@@ -379,7 +385,7 @@ def finalize_wasm(infile, outfile, memfile, DEBUG):
   building.save_intermediate(infile, 'base.wasm')
   # tell binaryen to look at the features section, and if there isn't one, to use MVP
   # (which matches what llvm+lld has given us)
```
Comment (tlively) [inline emscripten.py:387]: This comment looks like it is now out of date.

File: emscripten.py [hunk 2]
```diff
@@ -119,7 +119,13 @@ def update_settings_glue(metadata, DEBUG):
   all_exports = metadata['exports'] + list(metadata['namedGlobals'].keys())
   settings.WASM_EXPORTS = [asmjs_mangle(x) for x in all_exports]
 
-  settings.BINARYEN_FEATURES = metadata['features']
+  # start with the MVP features, and add any detected features.
+  settings.BINARYEN_FEATURES = ['--mvp-features'] + metadata['features']
+  if settings.USE_PTHREADS:
+    assert '--enable-threads' in settings.BINARYEN_FEATURES
+  if settings.MEMORY64:
+    assert '--enable-memory64' in settings.BINARYEN_FEATURES
```
Comment (tlively) [inline emscripten.py:127]: If we don't think these are (or should be) necessary, it would be good to add a TODO about that.

**Review by tlively — APPROVED** [review]
This looks like a nice simplification 👍  Was there a bug this fixes as well?

**Comment by sbc100** [comment]
> This looks like a nice simplification  Was there a bug this fixes as well?

No, there is no specific bug fix here.  It's more in preparation for a patch I have that transitions finalize itself to being optional.   In the current incarnation of that patch the features section is neither read nor stripped by finalize (because we don't run finalize at all).   I'm not sure that is the final place we want to land but this patch add flexibility which allows the features section to persist (or be extracted) at any point in the pipeline.

File: emscripten.py (hunk 2 above)
Comment (sbc100) [reply emscripten.py:127]: Yeah I mostly added this so the CI could prove they never fired.  I agree they are probably unnecessary.

**Pull request merged**